### DISCUSSION TITLE
First class support for computed values

### DISCUSSION THREAD

**Discussion by prenaissance** [pmndrs/zustand#1585] [Ideas] [answered]
I am aware of the middleware [zustand-middleware-computed-state](https://github.com/cmlarsen/zustand-middleware-computed-state) and other ways to select computed values, but I was wondering if the library would see use from first party computed values. The syntax can be leveraged from MobX, using getters, and the advantages would be simplicity to setup and readability.

```js
const useBearStore = create((set, get) => {
    bears: ["brown", "white"],
    selectedBearIndex: 0,
    get selectedBear() {
        return get().bears[selectedBearIndex]
    } // lazy loaded and cached
});
```

**Comment by dai-shi** [comment, marked as answer]
The primary solution is using selectors for computed values. But it only works in React.

We don't want to add code in core to support computed values, so middleware is the right solution. zustand is very simple without any tracking mechanism, so I wonder if mobx syntax is reasonable. (btw, have you considered [valtio](https://github.com/pmndrs/valtio)?)

fwiw, I'm experimenting a new approach: [derive-zustand](https://github.com/dai-shi/derive-zustand)
